docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+   ```
+
+* Device restart operation
+
+   ```
+   tedge/main/device/commands/req/device_restart
+   tedge/main/device/commands/res/device_restart
+   ```
+
+* Configuration snapshot operation
+
+   ```
+   tedge/main/service/<service-id>/commands/req/config_snapshot
+   tedge/main/service/<service-id>/commands/res/config_snapshot
+   ```
+
+* Configuration update operation
+
+   ```
+   tedge/<child-id>/service/<service-id>/commands/req/config_update
+   tedge/<child-id>/service/<service-id>/commands/res/config_update
+   ```
+
+Although all the above examples maintain consistent structure by ending with the `<operation-type>`,
+further additions are possible in future if desired for a given operation type.
+For e.g: `tedge/main/commands/req/config_update/<config-type>` to address a specific `config-type`
+
+### Registration service for child devices
+
+Immediate and nested child devices can be registered with thin-edge using its registration service,
+by sending the following MQTT message to the topic: `tedge/main/register/req/child`:
+
+```json
+{ 
+   "parent": "<parent-device-id>",
+   "id-prefix": "<desired-child-id-prefix>",
+   "capabilities": {
+      "<capability-1>": {},  //capability-1 specific metadata
+      "<capability-2>": {},  //capability-2 specific metadata
+      ...
+   }
+}
+```
+
+The `parent-device-id` is the device-id of the direct parent that the child device is connected to.
+The payload can have other fields describing the capabilities of that device as well (config management, software management etc).
+
+```admonish warning
+The exact topic keys and payload format for this init contract can be discussed and refined separately.
+The focus here is just on the MQTT topic structure.
+```
+
+Thin-edge needs to maintain the lineage (hierarchy of parent devices) of all descendent child devices in its internal state,
+so that it can be looked up while receiving any data from them.
+Even though the child device only declares its immediate parent in the registration message,
+the entire lineage can be traced back with a recursive lookup on that `parent-device-id`.
+
+The registration status, whether the device registration succeeded or not, is sent back on `tedge/main/register/res/child`,
+with the internal device id used by thin-edge to uniquely identify that device as follows:
+
+```json
+{
+   "id": "<generated-id>",
+   "status": "successful" 
+}
+```
+
+The `<generated-id>` will use the `<desired-child-id-prefix>` sent in the registration request.
+
+A failure is indicated with a failed status: `{ "status": "failed" }`.
+
+Once the registration is complete, these nested child devices can use the `tedge/<generated-id>` topic prefix
+to send telemetry data or receive commands as follows:
+
+
+|Type|Topic structure|
+|----|---------------|
+|Measurement|`tedge/descendent/<internal-id>/telemetry/measurements`|
+|Command|`tedge/descendent/<internal-id>/commands/req/software_list`|
```

Comment:
So the proposal is to use the `descendent` prefix when there is a registration step and `child` when not?

Or `descendent` grand-children and `child` for direct child?

Reply:
The latter: `descendent` for grand-children and `child` for direct child. Yes, `descendent` literally includes direct children as well. I used that word, more from the child's point-of-view than the parent ;-) Couldn't come up with something better. Suggestions welcome.

The distinction is there just so that `tedge` can decide whether to auto-register that entity or not. Any message coming form `child/` topics will go through auto-registration and the `descendent/` topics won't.

Reply:
Though shouldn't we also have the `device` and `service` in the prefixes, or is it assumed that the `<internal-id>` also includes the internal-id and the appropriate suffix (e.g. `/device` or `/service`)?

```suggestion
|Type|Topic structure|
|----|---------------|
|Measurement (device)|`tedge/descendent/<internal-id>/device/telemetry/measurements`|
|Command (device)|`tedge/descendent/<internal-id>/device/commands/req/software_list`|
|Measurement (service)|`tedge/descendent/<internal-id>/service/telemetry/measurements`|
|Command (service)|`tedge/descendent/<internal-id>/service/commands/req/software_list`|
```